x/gov/src/keeper.rs

```diff
@@ -314,9 +355,302 @@ impl<SK: StoreKey, PSK: ParamsSubspaceKey, M: Module, BK: BankKeeper<SK, M>>
+                ))
+            }
+        }
+
+        events
+    }
+
+    fn tally<DB: Database, CTX: TransactionalContext<DB, SK>>(
+        &self,
+        ctx: &mut CTX,
+        proposal_id: u64,
+    ) -> Result<(bool, bool, TallyResult), GasStoreErrors> {
+        let mut curr_validators = HashMap::<ValAddress, ValidatorGovInfo>::new();
+
+        for validator in self.staking_keeper.bonded_validators_by_power_iter(ctx)? {
+            let validator = validator?;
+
+            curr_validators.insert(
+                validator.operator().clone(),
+                ValidatorGovInfo {
+                    address: validator.operator().clone(),
+                    bounded_tokens: validator.bonded_tokens().clone(),
+                    delegator_shares: validator.delegator_shares().clone(),
+                    delegator_deduction: Decimal256::zero(),
+                    vote: Vec::new(),
+                },
+            );
+        }
+
+        let mut tally_results = TallyResultMap::new();
+        let mut total_voting_power = Decimal256::zero();
+
+        for vote in WeightedVoteIterator::new(ctx.kv_store(&self.store_key), proposal_id)
+            .map(|this| this.map(|(_, value)| value))
+            .collect::<Vec<_>>()
+        {
+            let MsgVoteWeighted {
+                proposal_id: _,
+                voter,
+                options: vote_options,
+            } = vote?;
+
+            let val_addr = ValAddress::from(voter.clone());
+            if let Some(validator) = curr_validators.get_mut(&val_addr) {
+                validator.vote = vote_options.clone();
+            }
+
+            for delegation in self
+                .staking_keeper
+                .delegations_iter(ctx, &voter)
+                .collect::<Vec<_>>()
+            {
+                let delegation = delegation?;
+
+                if let Some(validator) = curr_validators.get_mut(delegation.validator()) {
+                    // from cosmos: https://github.com/cosmos/cosmos-sdk/blob/d3f09c222243bb3da3464969f0366330dcb977a8/x/gov/keeper/tally.go#L51
+                    // There is no need to handle the special case that validator address equal to voter address.
+                    // Because voter's voting power will tally again even if there will deduct voter's voting power from validator.
+
+                    validator.delegator_deduction += delegation.shares(); // TODO: handle overflow?
+
+                    // delegation shares * bonded / total shares
+                    let voting_power = delegation
+                        .shares()
+                        .mul(Decimal256::new(validator.bounded_tokens))
+                        .div(validator.delegator_shares);
+
+                    for VoteOptionWeighted { option, weight } in &vote_options {
+                        let result_option = tally_results.get_mut(option);
+
+                        *result_option += voting_power * Decimal256::from(weight.clone());
```

Comment:
What is the type of `weight`?

Reply:
`Uint256`

Reply:
Maybe I missed something. It seems like it is incorrect conversion. Please, check it.

Reply:
I lied. It's wrapper around `Decimal256`. so no issue with it